### PR TITLE
New option in CSC DigiToRaw to pack strip digis by CFEB in each chamber. 

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

* New option in CSC DigiToRaw step to pack strip digis by CFEB in each chamber (called `packByCFEB`). This is how it was done in the (digital)CFEBs in Run-1 and Run-2 (and how it will be done in Run-3), but it was not implemented until now. The option is still turned off though.
* I also added a unit test to check if the DAQ conditions implemented in the CSC packer correctly pack the CSC wiregroups, comparators and strip digis. 10 dimuon events are generated with 0.9 < |eta| < 2.4 and 2 < pT < 1000 GeV. The whole simulation is done up to RawToDigi, followed by the test. In each event, check each chamber (540 CSC chambers) if the DAQ conditions are met. They are
   * Pack wire digis if there is an ALCT in that chamber (and a L1A)
   * Pack comparator digis if there is CLCT in that chamber (and a L1A)
   * Pack strip digis in a CFEB (strip sector in a chamber) if there is CLCT pretrigger in that CFEB (and a L1A)
* The DAQ conditions were moved to a static class `CSCDigiToRawAccept`, so that I can use the same functions in the unit test.

#### PR validation:

Because `packByCFEB` is set to `False` there should be no changes in any workflow.

Tested with WF 11634.0 

Also tested with `testPackingUnpackingPreTriggers_cfg.py`. To run the test, do `cmsRun EventFilter/CSCRawToDigi/test/testPackingUnpackingPreTriggers_cfg.py ` with either `packByCFEB` in cscPacker_cfi `True` or `False`. 

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

@barvic @ptcox @tahuang1991

### CODE REVIEW

**Comment by slava77** [comment]
@dildick 
please add a link to slides if this was already presented in a DPG meeting.

@cms-sw/csc-dpg-l2 
please clarify if we should expect a DPG code review here or if this was already checked before the PR was made.

Thank you.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by dildick** [comment]
@slava77: Tim Cox brought this up in a CSC DPG meeting some time ago. See https://indico.cern.ch/event/1056361/contributions/4443982/attachments/2277853/3869912/210707_csc_dpg_for_csc_meeting.pdf, page 8. It was on the CSC DPG wish list IIRC. Tim might be busy all week with CRUZET data certification though.

**Comment by dildick** [comment]
As expected, no differences for CSC in tests. 

@slava77 Should we reinitiate the tests after setting `packByCFEB` to True? We don't have to merge it in CMSSW that way, but it would help us understand the effects on RECO muons and PF candidate reconstruction and identification in MC simulation. And I could use that information in slides to present to CSC DPG.

**Comment by ptcox** [comment]
Victor and I will surely take a look at the code, and we would like Sven 
to advertise his work on this at a CSC meeting.

Slava Krutelyov wrote on 8/5/21 02:25:

**Comment by civanch** [comment]
+1

File: EventFilter/CSCRawToDigi/interface/CSCDigiToRawAccept.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ *  \author Sven Dildick - Rice
+ */
+
+#include "DataFormats/CSCDigi/interface/CSCALCTDigiCollection.h"
+#include "DataFormats/CSCDigi/interface/CSCCLCTDigiCollection.h"
+#include "DataFormats/CSCDigi/interface/CSCCLCTPreTriggerCollection.h"
+#include "DataFormats/CSCDigi/interface/CSCCLCTPreTriggerDigiCollection.h"
+
+class CSCDigiToRawAccept {
```
Comment (jpata) [inline EventFilter/CSCRawToDigi/interface/CSCDigiToRawAccept.h:16]: Is this class intended to store any state? What motivates the choice of a static class vs. the previous namespace?

File: EventFilter/CSCRawToDigi/src/CSCDigiToRawAccept.cc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  return result;
+}
+
+bool CSCDigiToRawAccept::accept(const CSCDetId& cscId,
+                                const CSCCLCTPreTriggerDigiCollection& lcts,
+                                int bxMin,
+                                int bxMax,
+                                int nominalBX,
+                                bool me1abCheck,
+                                std::vector<bool>& preTriggerInCFEB) {
```
Comment (jpata) [inline EventFilter/CSCRawToDigi/src/CSCDigiToRawAccept.cc:52]: there seems to be some code duplication here with respect to the other `accept` methods. Is there a possibility to define some common helper functions to reduce the duplication?

**Review by jpata — COMMENTED** [review]
> Should we reinitiate the tests after setting packByCFEB to True?

As far as I know, we typically don't use the bot tests for this type of validation of new features. For this feature, an offline validation of step3 in a suitable workflow could be sufficient.

I left some stylistic code comments, let's wait for the DPG's comments and a link to the presentation as well.

File: EventFilter/CSCRawToDigi/src/CSCDigiToRawAccept.cc (hunk 2 above)
Comment (dildick) [reply EventFilter/CSCRawToDigi/src/CSCDigiToRawAccept.cc:52]: There is no duplication. One function is for ALCT, CLCT and LCT. One function is for CLCTPreTriggerBX. One function is for CLCTPreTriggerDigi.

File: EventFilter/CSCRawToDigi/interface/CSCDigiToRawAccept.h (hunk 1 above)
Comment (dildick) [reply EventFilter/CSCRawToDigi/interface/CSCDigiToRawAccept.h:16]: I can replace it with a namespace if that's more desirable.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by dildick** [comment]
@jpata Here is the link to the presentation by CSC DPG https://indico.cern.ch/event/1056361/contributions/4443982/attachments/2277853/3869912/210707_csc_dpg_for_csc_meeting.pdf.

If there are no more comments on this PR, I propose that we merge it. (There are no changes in any workflows, because the option is not enabled yet). By mid September I'll try to run some tests, run the results by the CSC weekly meeting and then make a PR to enable the feature.

**Comment by jpata** [comment]
@cms-sw/csc-dpg-l2 did you have any additional comments?

**Comment by ptcox** [comment]
I haven’t had time to look at it yet. I’ll make time today.

**Comment by ptcox** [comment]
I've looked at the updated code and think it does exactly what is intended: models the real CSC readout suppression, and in particular lowers the strip readout granularity from by-chamber to by-CFEB, as in reality. Thanks, Sven. 
Speaking for CSC DPG we'd like this code released. As Sven notes, the suppression is OFF by default. We'll study it some more and at some point in the future request switching it ON.

**Comment by dildick** [comment]
I spotted a typo in the `CSCDigiToRawAccept` namespace

```C++
CSCDetId chamberId = chamberID(cscId);
  CSCCLCTPreTriggerDigiCollection::Range lctRange = lcts.get(chamberId);
  bool result = false;
  for (CSCCLCTPreTriggerDigiCollection::const_iterator lctItr = lctRange.first; lctItr != lctRange.second; ++lctItr) {
    int bx = lctItr->getBX() - nominalBX;
    if (bx >= bxMin && bx <= bxMax) {
      atLeastOnePreTrigger = true;
      // save the location of all pretriggers
      preTriggerInCFEB[lctItr->getCFEB()] = true;
    }
  }
  bool me1a = cscId.station() == 1 && cscId.ring() == 4;
  if (me1a && result == false && me1abCheck) {
    //check pretriggers in me1a as well; relevant for TMB emulator writing to separate detIds
    lctRange = lcts.get(cscId);
 ```

```result``` here should be ```atLeastOnePreTrigger```. I don't think it matters, because the block that checks for pretriggers in ME1/a is a remnant of older code where mistakenly triggers and pretrigger were made in ME1/a with ring == 4. 

Perhaps it makes more sense to remove those additional checks but I did not want to go that far in this PR.

**Comment by dildick** [comment]
@ptcox Is that something you wanted to change as well? 

The ```me1abCheck``` is obsolete since triggers and pretriggers in ME1/1 always have ring == 1. 

These blocks are obsolete as well
``` c++
    bool me1 = cscId.station() == 1 && cscId.ring() == 1;
    //this is another "creative" recovery of smart ME1A-ME1B TMB logic cases:
    //wire selective readout requires at least one (A)LCT in the full chamber
    if (me1 && result == false && me1abCheck) {
      CSCDetId me1aId = CSCDetId(chamberId.endcap(), chamberId.station(), 4, chamberId.chamber(), 0);
      lctRange = lcts.get(me1aId);
      for (typename LCTCollection::const_iterator lctItr = lctRange.first; lctItr != lctRange.second; ++lctItr) {
        int bx = lctItr->getBX() - nominalBX;
        if (bx >= bxMin && bx <= bxMax) {
          result = true;
          break;
        }
      }
    }
```

```c++
    bool me1a = cscId.station() == 1 && cscId.ring() == 4;
    if (me1a && result == false && me1abCheck) {
      //check pretriggers in me1a as well; relevant for TMB emulator writing to separate detIds
      lctRange = lcts.get(cscId);
      for (CSCCLCTPreTriggerCollection::const_iterator lctItr = lctRange.first; lctItr != lctRange.second; ++lctItr) {
        int bx = *lctItr - nominalBX;
        if (bx >= bxMin && bx <= bxMax) {
          result = true;
          break;
        }
      }
    }
```

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by ptcox** [comment]
Yes I'd be happy to have this removed. I think you introduced it, 
perhaps with Vadim, long ago?

Sven Dildick wrote on 8/10/21 19:39:

**Comment by jpata** [comment]
> Yes I'd be happy to have this removed.

@dildick I'll wait then until you update this PR before relaunching the tests.

**Comment by qliphy** [comment]
please test

**Comment by ptcox** [comment]
I think Sven is on vacation right now. But expected back next week.

Joosep Pata wrote on 8/19/21 15:10:

**Comment by dildick** [comment]
@ptcox I did not introduce the feature in the CSC packer. Slava did in 2015 after he realized that Vadim split up the ME1/1 readout in an ME1/a and ME1/b, which happened in 2012. I'll go ahead and remove it.

**Comment by dildick** [comment]
The last commit removes the me1abCheck. There should be no changes in any workflow.

**Comment by dildick** [comment]
@jpata Can you restart the tests, please?

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction
- new option for CSC strip digi packing (off by default)
- consequently, no reco changes expected or observed

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1
- new option for CSC strip digi packing, kept off by now but still testable with the included unit test

**Pull request merged**